### DISCUSSION TITLE
collaborative editing support with `yjs`

### DISCUSSION THREAD

**Discussion by RyKilleen** [josdejong/svelte-jsoneditor#192] [Ideas]
I was exploring solutions for collaborative editing experiences, and realized it might be a great pairing with the new svelte-json-editor! I think the support for JSON patches means it'd pair really well with [yjs](https://docs.yjs.dev/) ! It likely doesn't make sense as a core feature but as a supporting library, and I was curious if there's any preferred pattern for that.

I was wondering: how feasible it is to support multiple cursors in JSON mode? Implementation seems most straight-forward in text mode, but it might be cool to show multiple personalized cursors in JSON mode.

**Comment by josdejong** [comment]
Thanks for your suggestion. Collaborative editing would be really cool indeed, though I do not have a concrete use case for it myself right now. 

Some initial thoughts:

1. I made sure the API reports changes in small, atomic changes, using JSON Patch. That should indeed make it possible to do collaborative editing in the first place, thought there is more to it.
2. There is currently no public API to get/set the selection. That is necessary to be able to share your cursor. There is an open issue about that #163
3. Currently, `tree` mode supports only 1 selection. I think it would be doable to change this to support multiple selections, each with their own meta data like name and color.
4. Currently, `tree` mode does output JSON patch changes, but `text` mode not, that just replaces the full document. Now, `text` mode is powered by CodeMirror 6, and I see Codemirror is already supported by `yjs`, so that is probably a matter of opening up the API of CodeMirror.
5. I think the internal model of the `svelte-jsoneditor` needs a refactor to better group data, state, and changes together, and offer a unified API for the different modes (`code` and `text`). That needs some thinking through.
6. What if one user is using `text` mode, and the other is using `tree` mode? Do we want to unify the cursor/selection of both modes and be able to show the user in both modes? I'm not sure if that is possible, but having a unified API for both modes would be a big advantage.
